query-projector/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+};
+
+use mentat_db::{
+    SQLValueType,
+    TypedSQLValue,
+};
+
+use mentat_query::{
+    Element,
+    FindSpec,
+    PlainSymbol,
+    Variable,
+};
+
+use mentat_query_algebrizer::{
+    AlgebraicQuery,
+    DatomsColumn,
+    QualifiedAlias,
+    /*
+    ConjoiningClauses,
+    DatomsTable,
+    SourceAlias,
+    */
+};
+
+use mentat_query_sql::{
+    ColumnOrExpression,
+    /*
+    Constraint,
+    FromClause,
+    */
+    Name,
+    Projection,
+    ProjectedColumn,
+    /*
+    SelectQuery,
+    TableList,
+    */
+};
+
+error_chain! {
+    types {
+        Error, ErrorKind, ResultExt, Result;
+    }
+
+    foreign_links {
+        Rusqlite(rusqlite::Error);
+    }
+
+    links {
+        DbError(mentat_db::Error, mentat_db::ErrorKind);
+    }
+}
+
+#[derive(Debug)]
+pub enum QueryResults {
+    Scalar(Option<TypedValue>),
+    Tuple(Option<Vec<TypedValue>>),
+    Coll(Vec<TypedValue>),
+    Rel(Vec<Vec<TypedValue>>),
+}
+
+impl QueryResults {
+    pub fn len(&self) -> usize {
+        use QueryResults::*;
+        match self {
+            &Scalar(ref o) => if o.is_some() { 1 } else { 0 },
+            &Tuple(ref o)  => if o.is_some() { 1 } else { 0 },
+            &Coll(ref v)   => v.len(),
+            &Rel(ref v)    => v.len(),
+        }
+    }
+}
+
+type Index = i32;            // See rusqlite::RowIndex.
+type ValueTypeTag = i32;
+enum TypedIndex {
+    Known(Index, ValueTypeTag),
+    Unknown(Index, Index),
+}
+
+impl TypedIndex {
+    /// Look up this index and type(index) pair in the provided row.
+    /// This function will panic if:
+    ///
+    /// - This is an `Unknown` and the retrieved type code isn't an i32.
+    /// - If the retrieved value can't be coerced to a rusqlite `Value`.
+    /// - Either index is out of bounds.
+    ///
+    /// This function will return an error if the type code is unknown, or the value is
+    /// otherwise not convertible by the DB layer.
+    fn lookup<'a, 'stmt>(&self, row: &Row<'a, 'stmt>) -> Result<TypedValue> {
+        use TypedIndex::*;
+
+        match self {
+            &Known(value_index, value_type) => {
+                let v: rusqlite::types::Value = row.get(value_index);
+                TypedValue::from_sql_value_pair(v, value_type).map_err(|e| e.into())
+            },
+            &Unknown(value_index, type_index) => {
+                let v: rusqlite::types::Value = row.get(value_index);
+                let value_type_tag: i32 = row.get(type_index);
+                TypedValue::from_sql_value_pair(v, value_type_tag).map_err(|e| e.into())
+            },
+        }
+    }
+}
+
+// TODO: this sucks. Can we take ownership and avoid the clone?
+fn column_name(var: &Variable) -> Name {
+    let &Variable(PlainSymbol(ref s)) = var;
+    s.clone()
+}
+
+fn value_type_tag_name(var: &Variable) -> Name {
+    let &Variable(PlainSymbol(ref s)) = var;
+    format!("{}_value_type_tag", s)
+}
+
+/// Walk an iterator of `Element`s, collecting projector templates and
```

Comment:
I feel like you're making some assumptions about uniqueness of `Element` variable names, or some assumption about the SQL names they'll be mapped to.  What does the code do in the case of `:find ?x ?x :where ?x ...`?  Do you get two SQL variables, or just one?  (I expect two, given my reading of the patch.)